File: bot/alert_handlers.py
import logging
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import ContextTypes, CommandHandler, CallbackQueryHandler, ConversationHandler, MessageHandler, filters

from trading.alerts import add_price_alert, remove_price_alert, get_user_alerts
from database import get_user_watchlists, create_watchlist

# Setup logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# Conversation states
TOKEN_ADDRESS, PRICE_HIGH, PRICE_LOW, WATCHLIST_NAME, CONFIRM_ALERT = range(5)

async def alerts_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Handle the /alerts command."""
    user = update.effective_user
    
    # Send a "loading" message
    message = await update.message.reply_text("🔔 Loading your alerts...")
    
    # Get user's alerts
    alerts = await get_user_alerts(str(user.id))
    
    if "error" in alerts:
        await message.edit_text(f"❌ {alerts['message']}")
        return
    
    # Format alerts message
    alerts_text = f"🔔 *Your Price Alerts*\n\n"
    
    if alerts['alerts']:
        for alert in alerts['alerts']:
            alerts_text += f"• *{alert['token_symbol']}* - {alert['token_name']}\n"
            
            if alert['price_high']:
                alerts_text += f"  High: {alert['price_high']:.6f} ETH"
                if alert['current_price']:
                    alerts_text += f" (Current: {alert['current_price']:.6f} ETH)\n"
                else:
                    alerts_text += "\n"
            
            if alert['price_low']:
                alerts_text += f"  Low: {alert['price_low']:.6f} ETH"
                if alert['current_price']:
                    alerts_text += f" (Current: {alert['current_price']:.6f} ETH)\n"
                else:
                    alerts_text += "\n"
            
            alerts_text += f"  Address: `{alert['token_address']}`\n\n"
            
            # Store token address in context
            context.user_data[f"alert_token_{alert['token_address']}"] = alert
    else:
        alerts_text += "You don't have any price alerts set up yet.\n\nUse /addalert to add a new alert."
    
    # Create keyboard with remove buttons
    keyboard = []
    if alerts['alerts']:
        for alert in alerts['alerts']:
            keyboard.append([
                InlineKeyboardButton(
                    f"Remove {alert['token_symbol']} Alert", 
                    callback_data=f"remove_alert_{alert['token_address']}"
                )
            ])
    
    keyboard.append([InlineKeyboardButton("Add New Alert", callback_data="add_alert")])
    
    reply_markup = InlineKeyboardMarkup(keyboard)
    
    # Edit the "loading" message with the alerts information
    await message.edit_text(alerts_text, parse_mode='Markdown', reply_markup=reply_markup)

async def add_alert_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Handle the /addalert command."""
    # Check if token address is provided in arguments
    if context.args:
        context.user_data['alert_token_address'] = context.args[0]
        await update.message.reply_text(
            f"Setting up alert for token address: `{context.args[0]}`\n\n"
            f"Please enter the high price threshold in ETH (or 'skip' to skip):",
            parse_mode='Markdown'
        )
        return PRICE_HIGH
    
    # Ask for token address
    await update.message.reply_text(
        "Please enter the token address you want to set an alert for:"
    )
    return TOKEN_ADDRESS

async def token_address_input(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Handle token address input."""
    token_address = update.message.text
    
    # Validate token address (simple validation)
    if not token_address.startswith("0x") or len(token_address) != 42:
        await update.message.reply_text(
            "❌ Invalid token address. Please enter a valid Ethereum address:"
        )
        return TOKEN_ADDRESS
    
    # Store token address
    context.user_data['alert_token_address'] = token_address
    
    # Ask for high price threshold
    await update.message.reply_text(
        "Please enter the high price threshold in ETH (or 'skip' to skip):"
    )
    return PRICE_HIGH

async def price_high_input(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Handle high price threshold input."""
    price_high = update.message.text.lower()
    
    if price_high == "skip":
        context.user_data['alert_price_high'] = None
    else:
        try:
            price_high = float(price_high)
            if price_high <= 0:
                raise ValueError("Price must be greater than 0")
            
            context.user_data['alert_price_high'] = price_high
        except ValueError:
            await update.message.reply_text(
                "❌ Invalid price. Please enter a valid number or 'skip':"
            )
            return PRICE_HIGH
    
    # Ask for low price threshold
    await update.message.reply_text(
        "Please enter the low price threshold in ETH (or 'skip' to skip):"
    )
    return PRICE_LOW

async def price_low_input(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Handle low price threshold input."""
    price_low = update.message.text.lower()
    
    if price_low == "skip":
        context.user_data['alert_price_low'] = None
    else:
        try:
            price_low = float(price_low)
            if price_low <= 0:
                raise ValueError("Price must be greater than 0")
            
            context.user_data['alert_price_low'] = price_low
        except ValueError:
            await update.message.reply_text(
                "❌ Invalid price. Please enter a valid number or 'skip':"
            )
            return PRICE_LOW
    
    # Check if at least one threshold is set
    if context.user_data['alert_price_high'] is None and context.user_data['alert_price_low'] is None:
        await update.message.reply_text(
            "❌ You must set at least one price threshold. Please enter the high price threshold:"
        )
        return PRICE_HIGH
    
    # Get user's watchlists
    user = update.effective_user
    watchlists = get_user_watchlists(str(user.id))
    
    if watchlists:
        # Ask if user wants to add to watchlist
        watchlist_text = "Do you want to add this token to a watchlist? Choose one or type a new watchlist name:\n\n"
        
        keyboard = []
        for watchlist in watchlists:
            keyboard.append([InlineKeyboardButton(watchlist.name, callback_data=f"watchlist_{watchlist.name}")])
        
        keyboard.append([InlineKeyboardButton("No Watchlist", callback_data="watchlist_none")])
        keyboard.append([InlineKeyboardButton("Create New Watchlist", callback_data="watchlist_new")])
        
        reply_markup = InlineKeyboardMarkup(keyboard)
        
        await update.message.reply_text(watchlist_text, reply_markup=reply_markup)
        return WATCHLIST_NAME
    else:
        # No watchlists, ask if user wants to create one
        keyboard = [
            [InlineKeyboardButton("Yes, create watchlist", callback_data="watchlist_new")],
            [InlineKeyboardButton("No, skip watchlist", callback_data="watchlist_none")]
        ]
        
        reply_markup = InlineKeyboardMarkup(keyboard)
        
        await update.message.reply_text(
            "You don't have any watchlists yet. Would you like to create one?",
            reply_markup=reply_markup
        )
        return WATCHLIST_NAME

async def watchlist_button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Handle watchlist button presses."""
    query = update.callback_query
    await query.answer()
    
    data = query.data
    
    if data == "watchlist_none":
        # Skip watchlist
        context.user_data['alert_watchlist'] = None
        return await confirm_alert(update, context)
    
    elif data == "watchlist_new":
        # Ask for new watchlist name
        await query.edit_message_text(
            "Please enter a name for your new watchlist:"
        )
        return WATCHLIST_NAME
    
    elif data.startswith("watchlist_"):
        # Use existing watchlist
        watchlist_name = data[10:]
        context.user_data['alert_watchlist'] = watchlist_name
        return await confirm_alert(update, context)

async def watchlist_name_input(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Handle watchlist name input."""
    watchlist_name = update.message.text
    
    # Create new watchlist
    user = update.effective_user
    create_watchlist(str(user.id), watchlist_name)
    
    # Store watchlist name
    context.user_data['alert_watchlist'] = watchlist_name
    
    return await confirm_alert(update, context)

async def confirm_alert(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Confirm and create the alert."""
    # Get alert details
    token_address = context.user_data['alert_token_address']
    price_high = context.user_data['alert_price_high']
    price_low = context.user_data['alert_price_low']
    watchlist_name = context.user_data.get('alert_watchlist')
    
    # Create message
    if isinstance(update.callback_query, object):
        # If called from a button callback
        message_func = update.callback_query.edit_message_text
    else:
        # If called from a text message
        message_func = update.message.reply_text
    
    # Create the alert
    user = update.effective_user
    result = await add_price_alert(
        str(user.id),
        token_address,
        price_high,
        price_low,
        watchlist_name
    )
    
    if "error" in result:
        await message_func(f"❌ {result['message']}")
    else:
        confirm_text = f"✅ Alert added successfully!\n\n"
        confirm_text += f"*Token:* {result['token_symbol']} ({result['token_name']})\n"
        
        if price_high:
            confirm_text += f"*High Price:* {price_high} ETH\n"
        
        if price_low:
            confirm_text += f"*Low Price:* {price_low} ETH\n"
        
        if watchlist_name:
            confirm_text += f"*Watchlist:* {watchlist_name}\n"
        
        await message_func(confirm_text, parse_mode='Markdown')
    
    # Clear user data
    if 'alert_token_address' in context.user_data:
        del context.user_data['alert_token_address']
    
    if 'alert_price_high' in context.user_data:
        del context.user_data['alert_price_high']
    
    if 'alert_price_low' in context.user_data:
        del context.user_data['alert_price_low']
    
    if 'alert_watchlist' in context.user_data:
        del context.user_data['alert_watchlist']
    
    return ConversationHandler.END

async def alert_button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Handle alert button presses."""
    query = update.callback_query
    await query.answer()
    
    data = query.data
    
    if data == "add_alert":
        # Start add alert conversation
        await query.edit_message_text(
            "Please enter the token address you want to set an alert for:"
        )
        return TOKEN_ADDRESS
    
    elif data.startswith("remove_alert_"):
        # Remove alert
        token_address = data[13:]
        
        # Remove the alert
        user = update.effective_user
        result = await remove_price_alert(str(user.id), token_address)
        
        if "error" in result:
            await query.edit_message_text(f"❌ {result['message']}")
        else:
            await query.edit_message_text(f"✅ {result['message']}")

def register_alert_handlers(application):
    """Register alert handlers."""
    application.add_handler(CommandHandler("alerts", alerts_command))
    
    # Add alert conversation handler
    add_alert_conv_handler = ConversationHandler(
        entry_points=[
            CommandHandler("addalert", add_alert_command),
            CallbackQueryHandler(alert_button, pattern=r"^add_alert$")
        ],
        states={
            TOKEN_ADDRESS: [MessageHandler(filters.TEXT & ~filters.COMMAND, token_address_input)],
            PRICE_HIGH: [MessageHandler(filters.TEXT & ~filters.COMMAND, price_high_input)],
            PRICE_LOW: [MessageHandler(filters.TEXT & ~filters.COMMAND, price_low_input)],
            WATCHLIST_NAME: [
                CallbackQueryHandler(watchlist_button, pattern=r"^watchlist_"),
                MessageHandler(filters.TEXT & ~filters.COMMAND, watchlist_name_input)
            ],
            CONFIRM_ALERT: [MessageHandler(filters.TEXT & ~filters.COMMAND, confirm_alert)]
        },
        fallbacks=[CommandHandler("cancel", lambda u, c: ConversationHandler.END)]
    )
    application.add_handler(add_alert_conv_handler)
    
    # Alert button handler
    application.add_handler(CallbackQueryHandler(alert_button, pattern=r"^remove_alert_"))
    
    return application

File: src/bot/handlers/alertHandlers.js
/**
 * Alert Handlers for Zoracle Telegram Bot
 */
const { ethers } = require('ethers');
const { getTokenInfo } = require('../baseBot');
const { CONFIG } = require('../../config');

// In-memory alert storage (in a real implementation, this would be stored in a database)
const alerts = new Map();

// Alert states for conversation
const ALERT_STATES = {
  AWAITING_TOKEN: 1,
  AWAITING_PRICE_HIGH: 2,
  AWAITING_PRICE_LOW: 3
};

// In-memory alert state storage
const alertStates = new Map();

module.exports = (bot, users) => {
  // Alerts command
  bot.onText(/\/alerts/, async (msg) => {
    const chatId = msg.chat.id;
    const userId = msg.from.id.toString();
    
    // Check if user has a wallet
    const user = users.get(userId);
    
    if (!user || !user.wallet) {
      bot.sendMessage(chatId, 'You don\'t have a wallet set up yet. Use /wallet to set up a wallet.');
      return;
    }
    
    // Get user's alerts
    const userAlerts = alerts.get(userId) || [];
    
    if (userAlerts.length === 0) {
      const options = {
        reply_markup: {
          inline_keyboard: [
            [{ text: '➕ Add Alert', callback_data: 'add_alert' }]
          ]
        }
      };
      
      bot.sendMessage(chatId, 'You don\'t have any alerts set up yet.', options);
      return;
    }
    
    // Build alerts message
    let message = '🔔 *Your Alerts*\n\n';
    
    for (let i = 0; i < userAlerts.length; i++) {
      const alert = userAlerts[i];
      message += `*${i + 1}. ${alert.tokenInfo.name} (${alert.tokenInfo.symbol})*\n`;
      
      if (alert.priceHigh) {
        message += `Price Above: ${alert.priceHigh} ETH\n`;
      }
      
      if (alert.priceLow) {
        message += `Price Below: ${alert.priceLow} ETH\n`;
      }
      
      message += '\n';
    }
    
    // Add inline keyboard buttons for managing alerts
    const inlineKeyboard = [
      [{ text: '➕ Add Alert', callback_data: 'add_alert' }]
    ];
    
    // Add a remove button for each alert
    for (let i = 0; i < userAlerts.length; i++) {
      inlineKeyboard.push([{ text: `❌ Remove Alert ${i + 1}`, callback_data: `remove_alert_${i}` }]);
    }
    
    const options = {
      parse_mode: 'Markdown',
      reply_markup: {
        inline_keyboard: inlineKeyboard
      }
    };
    
    bot.sendMessage(chatId, message, options);
  });
  
  // Add alert command
  bot.onText(/\/addalert/, async (msg) => {
    const chatId = msg.chat.id;
    const userId = msg.from.id.toString();
    
    // Check if user has a wallet
    const user = users.get(userId);
    
    if (!user || !user.wallet) {
      bot.sendMessage(chatId, 'You don\'t have a wallet set up yet. Use /wallet to set up a wallet.');
      return;
    }
    
    // Start alert creation flow
    alertStates.set(userId, { state: ALERT_STATES.AWAITING_TOKEN });
    
    bot.sendMessage(chatId, 'Please enter the token address you want to set an alert for:');
  });
  
  // Add alert with token address
  bot.onText(/\/addalert (.+)/, async (msg, match) => {
    const chatId = msg.chat.id;
    const userId = msg.from.id.toString();
    const tokenAddress = match[1].trim();
    
    // Check if user has a wallet
    const user = users.get(userId);
    
    if (!user || !user.wallet) {
      bot.sendMessage(chatId, 'You don\'t have a wallet set up yet. Use /wallet to set up a wallet.');
      return;
    }
    
    // Validate token address
    if (!ethers.utils.isAddress(tokenAddress)) {
      bot.sendMessage(chatId, '❌ Invalid token address. Please provide a valid Ethereum address.');
      return;
    }
    
    try {
      // Get token info
      const tokenInfo = await getTokenInfo(tokenAddress);
      
      // Store alert state
      alertStates.set(userId, { 
        state: ALERT_STATES.AWAITING_PRICE_HIGH,
        tokenAddress,
        tokenInfo
      });
      
      bot.sendMessage(chatId, `Setting up alert for ${tokenInfo.name} (${tokenInfo.symbol})\n\nPlease enter the price (in ETH) to alert when the token goes ABOVE this value (or type "skip" to skip):`);
    } catch (error) {
      bot.sendMessage(chatId, `❌ Error: ${error.message}`);
    }
  });
  
  // Handle alert button
  bot.on('callback_query', async (callbackQuery) => {
    const action = callbackQuery.data;
    const msg = callbackQuery.message;
    const chatId = msg.chat.id;
    const userId = callbackQuery.from.id.toString();
    
    if (action === 'add_alert') {
      // Start alert creation flow
      alertStates.set(userId, { state: ALERT_STATES.AWAITING_TOKEN });
      
      bot.sendMessage(chatId, 'Please enter the token address you want to set an alert for:');
      bot.answerCallbackQuery(callbackQuery.id);
    } else if (action.startsWith('remove_alert_')) {
      const alertIndex = parseInt(action.split('_')[2]);
      
      // Get user's alerts
      const userAlerts = alerts.get(userId) || [];
      
      if (alertIndex >= 0 && alertIndex < userAlerts.length) {
        // Remove the alert
        const removedAlert = userAlerts.splice(alertIndex, 1)[0];
        
        // Update alerts
        alerts.set(userId, userAlerts);
        
        bot.answerCallbackQuery(callbackQuery.id, { 
          text: `Alert for ${removedAlert.tokenInfo.symbol} removed!`,
          show_alert: true
        });
        
        // Update the message
        if (userAlerts.length === 0) {
          const options = {
            reply_markup: {
              inline_keyboard: [
                [{ text: '➕ Add Alert', callback_data: 'add_alert' }]
              ]
            }
          };
          
          bot.editMessageText('You don\'t have any alerts set up yet.', {
            chat_id: chatId,
            message_id: msg.message_id,
            ...options
          });
        } else {
          // Rebuild alerts message
          let message = '🔔 *Your Alerts*\n\n';
          
          for (let i = 0; i < userAlerts.length; i++) {
            const alert = userAlerts[i];
            message += `*${i + 1}. ${alert.tokenInfo.name} (${alert.tokenInfo.symbol})*\n`;
            
            if (alert.priceHigh) {
              message += `Price Above: ${alert.priceHigh} ETH\n`;
            }
            
            if (alert.priceLow) {
              message += `Price Below: ${alert.priceLow} ETH\n`;
            }
            
            message += '\n';
          }
          
          // Add inline keyboard buttons for managing alerts
          const inlineKeyboard = [
            [{ text: '➕ Add Alert', callback_data: 'add_alert' }]
          ];
          
          // Add a remove button for each alert
          for (let i = 0; i < userAlerts.length; i++) {
            inlineKeyboard.push([{ text: `❌ Remove Alert ${i + 1}`, callback_data: `remove_alert_${i}` }]);
          }
          
          const options = {
            parse_mode: 'Markdown',
            reply_markup: {
              inline_keyboard: inlineKeyboard
            }
          };
          
          bot.editMessageText(message, {
            chat_id: chatId,
            message_id: msg.message_id,
            ...options
          });
        }
      } else {
        bot.answerCallbackQuery(callbackQuery.id, { text: 'Alert not found.' });
      }
    }
  });
  
  // Handle alert creation flow
  bot.on('message', async (msg) => {
    const chatId = msg.chat.id;
    const userId = msg.from.id.toString();
    const text = msg.text;
    
    // Check if we're in an alert creation flow
    const alertState = alertStates.get(userId);
    
    if (!alertState) return;
    
    if (alertState.state === ALERT_STATES.AWAITING_TOKEN) {
      // Validate token address
      if (!ethers.utils.isAddress(text)) {
        bot.sendMessage(chatId, '❌ Invalid token address. Please provide a valid Ethereum address.');
        return;
      }
      
      try {
        // Get token info
        const tokenInfo = await getTokenInfo(text);
        
        // Update alert state
        alertStates.set(userId, { 
          ...alertState,
          state: ALERT_STATES.AWAITING_PRICE_HIGH,
          tokenAddress: text,
          tokenInfo
        });
        
        bot.sendMessage(chatId, `Setting up alert for ${tokenInfo.name} (${tokenInfo.symbol})\n\nPlease enter the price (in ETH) to alert when the token goes ABOVE this value (or type "skip" to skip):`);
      } catch (error) {
        bot.sendMessage(chatId, `❌ Error: ${error.message}`);
      }
    } else if (alertState.state === ALERT_STATES.AWAITING_PRICE_HIGH) {
      let priceHigh = null;
      
      if (text.toLowerCase() !== 'skip') {
        // Parse price
        priceHigh = parseFloat(text);
        
        // Validate price
        if (isNaN(priceHigh) || priceHigh <= 0) {
          bot.sendMessage(chatId, '❌ Please enter a valid price or "skip".');
          return;
        }
      }
      
      // Update alert state
      alertStates.set(userId, { 
        ...alertState,
        state: ALERT_STATES.AWAITING_PRICE_LOW,
        priceHigh
      });
      
      bot.sendMessage(chatId, 'Please enter the price (in ETH) to alert when the token goes BELOW this value (or type "skip" to skip):');
    } else if (alertState.state === ALERT_STATES.AWAITING_PRICE_LOW) {
      let priceLow = null;
      
      if (text.toLowerCase() !== 'skip') {
        // Parse price
        priceLow = parseFloat(text);
        
        // Validate price
        if (isNaN(priceLow) || priceLow <= 0) {
          bot.sendMessage(chatId, '❌ Please enter a valid price or "skip".');
          return;
        }
      }
      
      // Validate that at least one price is set
      if (alertState.priceHigh === null && priceLow === null) {
        bot.sendMessage(chatId, '❌ You must set at least one price alert (high or low).');
        
        // Go back to price high
        alertStates.set(userId, { 
          ...alertState,
          state: ALERT_STATES.AWAITING_PRICE_HIGH
        });
        
        bot.sendMessage(chatId, 'Please enter the price (in ETH) to alert when the token goes ABOVE this value (or type "skip" to skip):');
        return;
      }
      
      // Create alert
      const alert = {
        tokenAddress: alertState.tokenAddress,
        tokenInfo: alertState.tokenInfo,
        priceHigh: alertState.priceHigh,
        priceLow,
        createdAt: Date.now()
      };
      
      // Get user's alerts
      const userAlerts = alerts.get(userId) || [];
      
      // Add alert
      userAlerts.push(alert);
      alerts.set(userId, userAlerts);
      
      // Clear alert state
      alertStates.delete(userId);
      
      // Build confirmation message
      let message = `✅ Alert created for ${alert.tokenInfo.name} (${alert.tokenInfo.symbol})!\n\n`;
      
      if (alert.priceHigh !== null) {
        message += `You will be notified when the price goes above ${alert.priceHigh} ETH.\n`;
      }
      
      if (alert.priceLow !== null) {
        message += `You will be notified when the price goes below ${alert.priceLow} ETH.\n`;
      }
      
      bot.sendMessage(chatId, message);
    }
  });
  
  // Simulate alert checking (in a real implementation, this would run on a schedule)
  setInterval(() => {
    // Check all alerts
    for (const [userId, userAlerts] of alerts.entries()) {
      for (const alert of userAlerts) {
        // Simulate price check (in a real implementation, we would fetch the actual price)
        const currentPrice = Math.random() * 0.5; // Random price between 0 and 0.5 ETH
        
        // Check if price triggers alert
        if (alert.priceHigh !== null && currentPrice > alert.priceHigh) {
          // Price went above threshold
          const message = `🔔 *Price Alert*\n\n${alert.tokenInfo.name} (${alert.tokenInfo.symbol}) price is now ${currentPrice.toFixed(6)} ETH, which is above your alert threshold of ${alert.priceHigh} ETH.`;
          
          bot.sendMessage(userId, message, { parse_mode: 'Markdown' });
        } else if (alert.priceLow !== null && currentPrice < alert.priceLow) {
          // Price went below threshold
          const message = `🔔 *Price Alert*\n\n${alert.tokenInfo.name} (${alert.tokenInfo.symbol}) price is now ${currentPrice.toFixed(6)} ETH, which is below your alert threshold of ${alert.priceLow} ETH.`;
          
          bot.sendMessage(userId, message, { parse_mode: 'Markdown' });
        }
      }
    }
  }, 60000); // Check every minute (in a real implementation, this would be more sophisticated)
}; 